pydantic_evals/pydantic_evals/datasets.py

```diff
@@ -0,0 +1,376 @@
+        return evaluation_rows
+
+    # TODO: Task: Always save a schema file when saving the dataset
+    def save(self, path: Path | str = DEFAULT_DATASET_PATH, schema_ref: str | None = None) -> None:
+        path = self._get_relative_path(path)
+        content = yaml.dump(to_jsonable_python(self), sort_keys=False)
+        if schema_ref is not None:
+            content = _ensure_yaml_language_server_line(content, schema_ref)
+        path.write_text(content)
+
+    @classmethod
+    def from_yaml(cls, path: Path | str = DEFAULT_DATASET_PATH) -> Self:
+        path = cls._get_relative_path(path)
+        if not path.exists():
+            raise FileNotFoundError(f'{cls.__name__} dataset file {path} does not exist')
+
+        raw = path.read_text()
+        loaded = yaml.safe_load(raw)
+        try:
+            result = cls.model_validate(loaded)
+            # result.rows = [cls.serialized_row_type().model_validate(row.model_dump()) for row in result.rows]
+        except ValidationError as e:
+            raise ValueError(
+                f'{cls.__name__} dataset file {path} contains data that does not match the schema:\n{e}.'
+            ) from e
+        return result
+
+    @classmethod
+    def generate_dataset_files(
+        cls,
+        dataset_path: Path | str = DEFAULT_DATASET_PATH,
+        schema_path: Path | str | None = None,
+    ) -> str:
+        dataset_path = cls._get_relative_path(dataset_path)
+
+        if schema_path is None:
+            if dataset_path.exists():
+                # Try to infer the schema path from the first line of the existing dataset file
+                first_line = dataset_path.read_text().split('\n', 1)[0]
+                if first_line.startswith('# yaml-language-server: $schema='):
+                    schema_path = (dataset_path.parent / first_line.split('=', 1)[1]).resolve()
+            if schema_path is None:
+                schema_path = dataset_path.parent / f'./{dataset_path.stem}_schema.json'
+        else:
+            schema_path = cls._get_relative_path(schema_path)
+
+        schema_content = to_json(cls.model_json_schema_with_assessments(), indent=2).decode() + '\n'
+        if not schema_path.exists() or schema_path.read_text() != schema_content:
+            schema_path.write_text(schema_content)
+
+        schema_ref = str(_get_relative_path_reference(schema_path, dataset_path.parent))
+        yaml_language_server_line = f'# yaml-language-server: $schema={schema_ref}'
+        if dataset_path.exists():
+            try:
+                cls.from_yaml(dataset_path)
+            except ValueError as e:
+                if isinstance(e.__cause__, ValidationError):
+                    raise ValueError(
+                        f'{cls.__name__} dataset file {dataset_path} already exists, but does not contain compatible data.'
+                        f' Fix or delete the file before calling this function:\n{e.__cause__}'
+                    ) from e.__cause__
```

Comment:
`from` doesn't replace the exception in the traceback, it just tweaks wording. combined with your current version of `from_yaml` this will show the validation error three times.